railties/lib/rails/generators/generated_attribute.rb

```diff
@@ -65,11 +65,13 @@ def parse(column_definition)
         end
 
         def dangerous_name?(name)
-          ActiveRecord::Base.dangerous_attribute_method?(name)
+          defined?(ActiveRecord::Base) &&
+            ActiveRecord::Base.dangerous_attribute_method?(name)
         end
 
         def valid_type?(type)
           DEFAULT_TYPES.include?(type.to_s) ||
+            !defined?(ActiveRecord::Base) ||
```

Comment:
Sorry even in `valid_type?` probably the original `defined?(ActiveRecord::Base) &&` is true.
If `ActiveRecord::Base` is not defined, this will allow any type except `DEFAULT_TYPES` (in other words, `valid_type?` never return `false`).

Reply:
| DEFAULT_TYPES.include?(type.to_s) | !defined?(ActiveRecord::Base) | ActiveRecord::Base.connection.valid_type? | result |
| ------------- |-------------| -----|-------|
| true  |         | | true | 
|false | true | | true | 
|false  | false | true | true |
|false  | false | false | false |

Reply:
It would only return false if `name` is not included in `DEFAULT_TYPES` and `ActiveRecord::Base` is defined and `ActiveRecord::Base.connection.valid_type?` returns `false`.

Reply:
I could also extract the `ActiveRecord` checks to something like `valid_type_in_activerecord?` to make things clearer?

Reply:
```ruby
def valid_type?(type)
   DEFAULT_TYPES.include?(type.to_s) &&
     !invalid_active_record_type?(type.to_s)
end

def invalid_active_record_type?(type)
  defined?(ActiveRecord::Base) &&
    !ActiveRecord::Base.connection.valid_type?(name)
end
```

Reply:
I was thinking that before #42311 we only allowed `DEFAULT_TYPES` as valid types (i.e. I thought we should not allow non `DEFAULT_TYPES` if `ActiveRecord::Base` is not defined), but that is not true, before #42311 we didn't check type, so the matrix looks good to me. Thank you for your detailed explanation.

Reply:
Thanks @kamipo ! 🙇